From eab31108581a20f6909e6a802962e5c43360e28e Mon Sep 17 00:00:00 2001
From: Alexey Marchuk <alexeymar@mellanox.com>
Date: Thu, 15 Jul 2021 18:39:27 +0300
Subject: [PATCH 4/7] meson/mlx5: Suppress -Wunused-value diagnostic

mlx5 common library checks if several symbols/definitions
are presented in system header files. If some are not
presented, they will be enabled by mlx5_glue library.
The problem appears with clang and '-Werror' - code
generated by meson is not compiled due to unused variable:

Code:

        #include <infiniband/mlx5dv.h>
        int main(void) {
            /* If it's not defined as a macro, try to use as a symbol */
            #ifndef mlx5dv_create_flow_action_packet_reformat
                mlx5dv_create_flow_action_packet_reformat;
            #endif
            return 0;
        }
Compiler stdout:

Compiler stderr:
 /hpc/local/work/alexeymar/repo/spdk/dpdk/build-tmp/meson-private/tmp5obnak86/testfile.c:6:17: error: expression result unused [-Werror,-Wunused-value]
                mlx5dv_create_flow_action_packet_reformat;
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As result, almost all symbols are enabled in mlx5_glue while
they exist is system headers. As result, we get multiple
symbols redefenitions when we compile mlx5_common.
As a solution for this problem we can suppress
-Wunused-vaurable using pragma

Change-Id: I03ba5d03f7e53d8e593a9de1deace5140c67d21d
Signed-off-by: Alexey Marchuk <alexeymar@mellanox.com>
Signed-off-by: Krzysztof Karas <krzysztof.karas@intel.com>
Reviewed-on: https://review.spdk.io/gerrit/c/spdk/dpdk/+/8795 (spdk-21.05)
Reviewed-on: https://review.spdk.io/gerrit/c/spdk/dpdk/+/9219 (spdk-21.08)
Reviewed-on: https://review.spdk.io/gerrit/c/spdk/dpdk/+/10451 (spdk-21.11)
Reviewed-on: https://review.spdk.io/gerrit/c/spdk/dpdk/+/12499
(spdk-21.11.1)
Reviewed-on: https://review.spdk.io/gerrit/c/spdk/dpdk/+/14310
Tested-by: SPDK CI Jenkins <sys_sgci@intel.com>
Reviewed-by: Konrad Sztyber <konrad.sztyber@intel.com>
Reviewed-by: Ben Walker <benjamin.walker@intel.com>
---
 drivers/common/mlx5/linux/meson.build | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/common/mlx5/linux/meson.build b/drivers/common/mlx5/linux/meson.build
index 4c7b53b9bd..f5a4daeb31 100644
--- a/drivers/common/mlx5/linux/meson.build
+++ b/drivers/common/mlx5/linux/meson.build
@@ -205,7 +205,8 @@ has_sym_args = [
 ]
 config = configuration_data()
 foreach arg:has_sym_args
-    config.set(arg[0], cc.has_header_symbol(arg[1], arg[2], dependencies: libs))
+    file_prefix = '#pragma clang diagnostic ignored "-Wunused-value"'
+    config.set(arg[0], cc.has_header_symbol(arg[1], arg[2], prefix : file_prefix, dependencies: libs))
 endforeach
 foreach arg:has_member_args
     file_prefix = '#include <' + arg[1] + '>'
-- 
2.31.1

